File: Visual Studio 2008/VBEFEntity/VBEFEntityDataModel/One2Many/One2ManyClass.vb
'****************************** 模块头 ******************************'
' 模块名:        One2ManyClass.vb
' 项目名:        VBEFEntityDataModel
' 版权 (c)       Microsoft Corporation.
'
' 本例阐明了如何通过一对多关联插入，更新，和查询这两个实体
' 
' This source is subject to the Microsoft Public License.
' See http://www.microsoft.com/opensource/licenses.mspx#Ms-PL.
' All other rights reserved.
' 
' THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
' EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
' WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
'***************************************************************************'


#Region "Imports directives"
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
#End Region


Namespace VBEFEntityDataModel.One2Many
    Friend Class One2ManyClass

        ' 测试One2ManyClass中的所有方法
        Public Shared Sub One2ManyTest()
            InsertDepartmentWithCourse()

            InsertCourse()

            UpdateCourse()

            UpdateDepartment()

        End Sub

        ' 插入新的department包括新的course
        Public Shared Sub InsertDepartmentWithCourse()

            Using context As New EFO2MEntities()

                Dim department As New Department() With _
                { _
                    .DepartmentID = 4, _
                    .Name = "Software Engineering" _
                }

                Dim course As New Course() With _
                { _
                    .CourseID = 2202, _
                    .Title = "ADO.NET" _
                }

                department.Course.Add(course)

                context.AddToDepartment(department)

                Try

                    Console.WriteLine("Inserting department {0} with course " _
                        + "{1}", department.Name, course.Title)

                    context.SaveChanges()

                    Query()

                Catch ex As Exception
                    Console.WriteLine(ex.Message)
                End Try
            End Using
        End Sub


        ' 插入新的course到现存的department
        Public Shared Sub InsertCourse()

            Using context As New EFO2MEntities()
                Dim course As New Course() With _
                { _
                    .CourseID = 2203, _
                    .Title = "Object Oriented Programming" _
                }

                course.Department = (From p In context.Department _
                    Where p.DepartmentID = 7 _
                    Select p).First()

                context.AddToCourse(course)

                Try

                    Console.WriteLine("Inserting course {0} to department " _
                        + "{1}", course.Title, course.Department.Name)

                    context.SaveChanges()

                    Query()

                Catch ex As Exception
                    Console.WriteLine(ex.Message)
                End Try
            End Using
        End Sub


        ' 获取带有courses的所有departments 
        Public Shared Sub Query()

            Using context As New EFO2MEntities()

                Dim query = From p In context.Department.Include("Course") _
                            Select p

                Console.WriteLine("Deparments with their courses:")

                For Each d As Department In query

                    Console.WriteLine("{0} {1}", d.DepartmentID, d.Name)

                    For Each c As Course In d.Course

                        Console.WriteLine("  {0}", c.Title)

                    Next
                Next

                Console.WriteLine()

            End Using
        End Sub


        ' 更新一个现存的course
        Public Shared Sub UpdateCourse()

            Using context As New EFO2MEntities()

                Dim course As New Course()

                course.CourseID = 2203

                context.AttachTo("Course", course)

                course.Title = "OOP"

                Try

                    Console.WriteLine("Modifying Course 2203's Title to {0}", _
                        course.Title)

                    context.SaveChanges()

                    Query()

                Catch ex As Exception
                    Console.WriteLine(ex.Message)
                End Try
            End Using
        End Sub


        ' 更新一个现存的department 
        Public Shared Sub UpdateDepartment()

            Using context As New EFO2MEntities()

                Dim department As New Department()

                department.DepartmentID = 1

                context.AttachTo("Department", department)

                department.Name = "Computer Engineering"

                department.Course.Add(New Course() With _
                { _
                    .CourseID = 2204, _
                    .Title = "Arithmetic" _
                })

                Try

                    Console.WriteLine("Modifying Department 1's Title to {0}" _
                       + ", and insert a new Course 2204 into the " + _
                       "Department 1", department.Name)

                    context.SaveChanges()

                    Query()

                Catch ex As Exception
                    Console.WriteLine(ex.Message)
                End Try
            End Using
        End Sub

    End Class

End Namespace

File: Visual Studio 2008/CSEFEntity/CSEFEntityDataModel/One2Many/One2ManyClass.cs
/****************************** 模块头 ******************************\
* 模块名:        One2ManyClass.cs
* 项目名:        CSEFEntityDataModel
* 版权 (c)       Microsoft Corporation.
*
* 本例阐明了如何通过一对多关联插入，更新，和查询这两个实体
* 
* This source is subject to the Microsoft Public License.
* See http://www.microsoft.com/opensource/licenses.mspx#Ms-PL.
* All other rights reserved.
* 
* THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
* EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
* WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
\***************************************************************************/

#region Using directives
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
#endregion


namespace CSEFEntityDataModel.One2Many
{
    public static class One2ManyClass
    {
        // 测试One2ManyClass中的所有方法
        public static void One2ManyTest()
        {
            InsertDepartmentWithCourse();

            InsertCourse();

            UpdateCourse();

            UpdateDepartment();
        }


        // 插入新的department包括新的course
        public static void InsertDepartmentWithCourse()
        {
            using (EFO2MEntities context = new EFO2MEntities())
            {
                Department department = new Department() 
                { 
                    DepartmentID = 4, 
                    Name = "Software Engineering" 
                };

                Course course = new Course() 
                { 
                    CourseID = 2202, 
                    Title = "ADO.NET" 
                };

                department.Course.Add(course);

                context.AddToDepartment(department);

                try
                {
                    Console.WriteLine("Inserting department {0} with course "
                        + "{1}", department.Name, course.Title);

                    context.SaveChanges();

                    Query();
                }
                catch (Exception ex)
                {
                    Console.WriteLine(ex.Message);
                }
            }
        }


        // 插入新的course到现存的department
        public static void InsertCourse()
        {
            using (EFO2MEntities context = new EFO2MEntities())
            {
                Course course = new Course() 
                { 
                    CourseID = 2203, 
                    Title = "Object Oriented Programming" 
                };

                course.Department = (
                    from p in context.Department 
                    where p.DepartmentID == 7 
                    select p).First();

                context.AddToCourse(course);

                try
                {
                    Console.WriteLine("Inserting course {0} to department "
                        + "{1}", course.Title, course.Department.Name);

                    context.SaveChanges();

                    Query();
                }
                catch (Exception ex)
                {
                    Console.WriteLine(ex.Message);
                }
            }
        }


        // 获取带有courses的所有departments
        public static void Query()
        {
            using (EFO2MEntities context = new EFO2MEntities())
            {
                var query = from p in context.Department.Include("Course") 
                            select p;

                Console.WriteLine("Deparments with their courses:");
                foreach (Department d in query)
                {
                    Console.WriteLine("{0} {1}", d.DepartmentID, d.Name);

                    foreach (Course c in d.Course)
                    {
                        Console.WriteLine("   {0}", c.Title);
                    }
                }

                Console.WriteLine();
            }
        }


        // 更新一个现存的course
        public static void UpdateCourse()
        {
            using (EFO2MEntities context = new EFO2MEntities())
            {
                Course course = new Course();

                course.CourseID = 2203;

                context.AttachTo("Course", course);

                course.Title = "OOP";

                try
                {
                    Console.WriteLine("Modifying Course 2203's Title to {0}", 
                        course.Title);

                    context.SaveChanges();

                    Query();
                }
                catch (Exception ex)
                {
                    Console.WriteLine(ex.Message);
                }
            }
        }


        // 更新一个现存的department 
        public static void UpdateDepartment()
        {
            using (EFO2MEntities context = new EFO2MEntities())
            {
                Department department = new Department();

                department.DepartmentID = 1;

                context.AttachTo("Department", department);

                department.Name = "Computer Engineering";

                department.Course.Add(new Course() 
                { 
                    CourseID = 2204, 
                    Title = "Arithmetic" 
                });

                try
                {
                    Console.WriteLine("Modifying Department 1's Title to {0}"
                        + ", and insert a new Course 2204 into the " + 
                        "Department 1", department.Name);

                    context.SaveChanges();

                    Query();
                }
                catch (Exception ex)
                {
                    Console.WriteLine(ex.Message);
                }
            }
        }
    }
}
